### PR TITLE
only close new ports upon iptables-restore failure

### LINKED ISSUE #24357: iptables kube-proxy should not close all ports of nodeport service on iptables-restore failure
The symptom is the nodeport service is accessible thru hostip:hostport. But the port on host is not occupied. This is exposed by `Services should be able to create a functioning NodePort service` e2e on kubenet. 

I noticed in the kube-proxy.log. Kube-proxy did open the port. 
```
I0413 17:25:30.432151       5 proxier.go:1114] Opened local port "nodePort for e2e-tests-services-m5ern/nodeport-test:" (:31857/tcp)

```

But later, kube-proxy closed the ports upon iptables-restore failure. It never reopen the ports. Hence causing the e2e failure.
```
E0413 17:26:42.537617       5 proxier.go:941] Failed to execute iptables-restore: exit status 1 (iptables-restore: line 139 failed
)
E0413 17:26:42.537674       5 proxier.go:944] Closing local port "nodePort for kube-system/default-http-backend:http" (:30867/tcp)
E0413 17:26:42.537713       5 proxier.go:944] Closing local port "nodePort for e2e-tests-services-m5ern/nodeport-test:" (:31857/tcp)

```

The code section in question:  https://github.com/kubernetes/kubernetes/blob/release-1.2/pkg/proxy/iptables/proxier.go#L943

Not sure why `iptables-restore` failed. Error messge `line 139 failed` is not helpful since line 139 is `COMMIT`

ref: #24211

### PR DESCRIPTION
**Pull request by freehan**
fixes: #24357

### CODE REVIEW

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -941,8 +941,11 @@ func (proxier *Proxier) syncProxyRules() {
 		glog.Errorf("Failed to execute iptables-restore: %v", err)
 		// Revert new local ports.
 		for k, v := range newLocalPorts {
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: why does this map contain old ports to begin with?

Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: It is poorly named.  It represents the set of local ports "after the current state is synced".  It's "new" as in "old set vs new set" not as in "previously not opened".

Minhan, can you

* rename this variable to something like "finalPortsMap" or "replacementPortsMap"
* change the comment on it to say "Accumulate the set of local ports that we will be holding open once this update is complete"
* add comments at L733 and L796 to the effect of "this port was open before and is still needed"
* change the comment in your fix to something like "Only close *newly opened* local ports - leave ones that were open before this update"
* change the glog.Errorf to V(2).Infof and make the message clearer.  Something like "Not holding port %s open after failure"

If we want to get really fancy, maybe we want (as a different PR) to parse the iptables error looking for the exact error message, recognize any error with a line number, and print `lines`.  Even better, make a function that prints it with line numbers.

**Comment by ArtfulCoder** [comment]
@bprashanth  the map contains old ports to begin with because, after services are updated, the updated services might still be using ports that were in use by before-the-update-services.

**Comment by freehan** [comment]
ready round 2

**Comment by freehan** [comment]
ping

**Comment by thockin** [comment]
Change looks good - how about a test?  Maybe we can break the cleanup logic off into a function that takes the existing map and the replacement map and closes "new" ports, then you can unit test that function with fake `closeable` objects and ensure the right set was closed.

We really need to strive that every PR increases coverage or at least doesn't DECREASE it

**Comment by freehan** [comment]
Alright. Test added.

**Comment by freehan** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by freehan** [comment]
@k8s-bot unit test this issue: #IGNORE

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -1198,3 +1197,15 @@ func openLocalPort(lp *localPort) (closeable, error) {
 	glog.V(2).Infof("Opened local port %s", lp.String())
 	return socket, nil
 }
+
+// revertPorts is closing ports in replacementPortsMap but not in existingPortsMap. In other words, it only
+// closes the ports opened in this sync.
+func revertPorts(replacementPortsMap, existingPortsMap map[localPort]closeable){
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: super nit: name existingPortsMap to originalPortsMap.  Clearer to read.

**Comment by thockin** [comment]
Just a name nit and I am happy.  Please self-LGTM, since I will be OOO this week.

**Comment by freehan** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by freehan** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by freehan** [comment]
@k8s-bot test this issue: #23974

**Pull request merged**